### ISSUE TITLE
call_fixture_func should test the return value not the function

### ISSUE THREAD

**Issue by mgrazebrook** [pytest-dev/pytest#7235]
I'm trying to use fixtures in a library-like context. It doesn't seem you can have a pip installed fixture so the fixture functions are in the library (less decorator) then decorated in the local conftest.py

This means that a library 'fixture' with clean-up code (eg. resetting a database state) has a yield but the decorated function does not.

fixtures.call_fixture_func tests if the decorated function is a generator. In a certain sense, the function itself is not a generator but behaves as a generator since it returns a generator. So could you test the return value not the function?

```python
import inspect

def f():
  print('before')
  yield 27
  print('after')

def g():
  return f()

print('inspect f:', inspect.isgeneratorfunction(f))
print('inspect g:', inspect.isgeneratorfunction(g))

for i, x in enumerate(f()):
    print(i, x)

ret = f()
print('is ret a generator?', inspect.isgenerator(ret))

print('a', next(ret))
try:
  print('b', next(ret))
except StopIteration:
  print('it worked')
```
Output:
```
$ python3.7 x.py
inspect f: True
inspect g: False
before
0 27
after
is ret a generator? True
before
a 27
after
it worked
```

pip list:

[piplist.txt](https://github.com/pytest-dev/pytest/files/4662130/piplist.txt)

OS version:
ProductName:	Mac OS X
ProductVersion:	10.14.6
BuildVersion:	18G103

**Comment by nicoddemus** [comment]
Hi @mgrazebrook,

Sorry it is not very clear to me what you want to accomplish... do you want to change pytest itself or be able to run generators as test functions?

Can you provide an example code and what you expect the pytest output to be with that code?

Thanks!

**Comment by RonnyPfannschmidt** [comment]
fixture implementations can happen in terms of other context managers

as far as i understood OP wants to run the return vaule of a incomplete contextmanager to work as if the fixture was the context manager

from my pov thats structurally wrong

instead of 
```
def f():
  yield

@fixture  
def g():
  return f
```
one should use something like
 
```
@contextmanager
def f():
  yield

@fixture  
def g():
  with f() as g:
    yield g
```
the second example has much clearer flow control, can't accidentally misinterpret a return value and uses standard python ideoms

i believe this issue can be closed as wont-fix

**Comment by mgrazebrook** [comment]
Given f and g as I defined them, returning the same value in either case, it struck me as odd pytest behaved differently and it wasn't until I debugged into the pytest code that I found out why.

Ronny makes a reasonable point that the slightly longer form he proposes is more ideomatic. 

On the other hand, my version says to me 'the function is wrapped as a fixture but otherwise unchanged' while writing it as nested context managers is - for me - more complex. So if simpler is  more pythonic, perhaps less ideomatic but more pythonic?

I came at  this from the point of view of building a library for which could be pip installed and thus shared between several programs. Initially I was stuck on how to re-use them. Pytest doesn't find them, though I did eventually discover that if I imported the functions rather than modules into conftest.py, I could have decorated and complete fixtures in the library and still re-use them. This wasn't at all clear t me from the documentation: it took me a long time to discover that.

So I'd attempted to work around the problem as I've documented it. The approach isn't as valuable as it seemed before I found I could import the fixtures but does still seem valid. It took a fair bit of debugging to work out why pytest worked in a way which was, to me, unexpected: if I return the same thing, I'd expect it to behave the same way.

So I can see an argument for saying this would be very low priority. Yet I also feel that the pytest code would be more correct - and perhaps save others in my position time - if you tested the return value not the function. After all, you call the fixture in both paths. Before you reject this, ask yourself if you thing pytest behaved in an expected way given the (ab-)use I put it to!

PS: The ability to have a fixture library and import fixtures is useful and perhaps worth documenting? Especially since there are some differences from normal imports such as the need to also import fixtures which are part of the re-used fixture's internal interface. My core goal is to make reusable fixture libraries.

**Comment by RonnyPfannschmidt** [comment]
Declared and transformed context mangers are not the same as functions that return something, 

Calling declared fixtures as functions is already deprecated and will eventually be removed.

The call contract for fixtures is: treat as context-manager when it contains a yield, else treat it as a function. We would actually break other peoples code if we suddenly changed that to "if it returns a generator, run it like it was a function context manager" . So from my pov this isn't low prio, this is "a breaking change we won't do"

Importing fixtures is a anti-pattern, as each import site counts as distinct fixture, fixture collections should be put into plugins.

Note that you can still use yield from to do actual deferring of a generator.

**Comment by mgrazebrook** [comment]
I'm not trying to declare a fixture and wrap it a second time, rather to share the 'fixture' undecorated by @fixture, turning it into a fixture in the application specific code.
To me, if it returns a generator, it is a generator function - even if the generator is created deeper in the code. 
So yes, what I propose is a change and thus could break existing code, as changes can. However if it did, it would probably be a good thing as it's hard to imagine a case where the pytest code as is, (ab)used as I did, would be behaving as the author intended. The only case I can imagine would be if someone used this as a work-around to create a fixture which returns a generator or context manager but is not treated as a context manager by pytest - but that would make for very confusing code.
However you've more experience than me and have made up your mind, so I accept that. Also this doesn't seem important since the use case which led me to this is covered by plugins. Now I've discovered their existence, my problem is solved.
While I still think the change I propose is an improvement, it's fine if you reject it and close the ticket.
